### DISCUSSION TITLE
parse error.log

### DISCUSSION THREAD

**Discussion by 63110** [martin-helmich/prometheus-nginxlog-exporter#311] [General]
Hi. 
I wanted to get log from error.log, but I can't to find format log. 
I know that, we can't to change log_format for error.log only level of errors. But which log_format use to parse? 
thanks for any suggestions
I read this article https://github.com/martin-helmich/prometheus-nginxlog-exporter/discussions/185 but it's not work for me.
Can't see fully messages

namespace "endpoint" {
  source = {
    files = [
      "/var/log/nginx/endpoint.log"
    ]
}
format = "$date $time [$level] $pid#$tid: *$cid $message"
  relabel "level" { from = "level" }
}
result 

endpoint_http_response_count_total{level="alert",method="",status=""} 1

**Comment by martin-helmich** [comment]
How does your error log look like? Does it match the format `"$date $time [$level] $pid#$tid: *$cid $message"`, or does it look like a different format entirely?

**Reply by 63110** [reply]
Yeah, My error.log 

2023/02/06 09:08:24 [alert] 105326#105326: *149 shared connection is busy while proxying and sending to client, udp client: 111.120.11.00, server: 0.0.0.0:17, upstream: "00.00.00.94:80", bytes from/to client:15820296/356453044, bytes from/to upstream:356454496/15820296
<img width="611" alt="Screenshot_14" src="https://user-images.githubusercontent.com/70537871/217752452-f1e34737-575e-48cb-ab3c-39ea29f04b7e.png">

**Reply by martin-helmich** [reply]
But, from your metrics, it looks like the errors are actually scraped successfully, and it even got the error level correct (apart from the metric being misleadingly named `http_response_count`, because that's just hardcoded in the exporter).

If you want to expose the `$message` as label, you'll need to add a relabeling config. Be advised that this may result in a high metrics churn rate[^1], leading in turn to a high number of different metrics and potentially increased memory consumption. In that case, I'd also suggest adding additional capturing groups to your log format to catch all the connection metadata that are present at the end of the line (so that they're not caught by the `$message` group).

[^1]: https://valyala.medium.com/prometheus-storage-technical-terms-for-humans-4ab4de6c3d48

**Reply by 63110** [reply]
Ohh. Now I stuck with problem, that delimiter is gap. Which don't allow to parse full message 
<img width="560" alt="Screenshot_22" src="https://user-images.githubusercontent.com/70537871/218428516-a066ab5b-e2d5-4c81-8f1e-061822162367.png">
but message field
*149 shared connection is busy while proxying and sending to client, udp client: 111.120.11.00, server: 0.0.0.0:17, upstream: "00.00.00.94:80", bytes from/to client:15820296/356453044, bytes from/to upstream:356454496/15820296.


P.s I tried to use  split and separator. but the same situation

**Reply by martin-helmich** [reply]
The regular expressions that are generated from the access log format are non-greedy by default. You should (untested!) be able to enforce matching the whole message by adding additional matchers after the `$message` -- even a generic `$remainer` matcher at the end should suffice, as long as it is unambiguous:

```
format = "$date $time [$level] $pid#$tid: *$cid $message, $protocol client: $client, server: $server, upstream: \"$upstream\", $remainder"
```

Also, on a side node:

- Do not use high-cardinality values (like process/connection IDs, timestamps, and so on) as label values in a Prometheus metric. Prometheus will store all different combination of label values as separate time series, resulting in a *massive* number of time series, which in turn will degrade performance and increase your memory consumption proportionally.

**Reply by 63110** [reply]
without changes...
does not match given format...
Can I somehow parse it differently for verification?
